File: app/src/main/java/org/baiyu/fuckshare/exifhelper/jpegExifHelper.java
package org.baiyu.fuckshare.exifhelper;

import android.os.FileUtils;
import android.util.Log;

import org.baiyu.fuckshare.Utils;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Set;

public class jpegExifHelper implements ExifHelper {

    private static final Set<Byte> jpegSkippableChunks = Set.of(
            (byte) 0xE0,   //
            (byte) 0xE1,   // exif, xmp, xap
            (byte) 0xE2,   // icc
            (byte) 0xE3,   // Kodak
            (byte) 0xE4,   // FlashPix
            (byte) 0xE5,   // Ricoh
            (byte) 0xE6,   // GoPro
            (byte) 0xE7,   // Pentax/Qualcomm
            (byte) 0xE8,   // Spiff
            (byte) 0xE9,   // MediaJukebox
            (byte) 0xEA,   // PhotoStudio
            (byte) 0xEB,   // HDR
            (byte) 0xEC,   // photoshoP ducky / savE foR web
            (byte) 0xED,   // photoshoP savE As
            (byte) 0xEE,   // "adobe" (length = 12)
            (byte) 0xEF,   // GraphicConverter
            (byte) 0xFE    // Comments
    );

    @Override
    public void removeMetadata(InputStream inputStream, OutputStream outputStream) {
        try {
            BufferedInputStream bis;
            BufferedOutputStream bos;

            if (inputStream instanceof BufferedInputStream) {
                bis = (BufferedInputStream) inputStream;
            } else {
                bis = new BufferedInputStream(inputStream);
            }

            if (outputStream instanceof BufferedOutputStream) {
                bos = (BufferedOutputStream) outputStream;
            } else {
                bos = new BufferedOutputStream(outputStream);
            }

            byte[] maker = new byte[2];
            byte[] lenBytes = new byte[2];
            long len;

            while (bis.available() > 0) {
                Utils.inputStreamRead(bis, maker);
                assert maker[0] == (byte) 0xFF;

                if (maker[1] == (byte) 0xD8) {
                    bos.write(maker);
                    Log.d("fuckshare", String.format("Chunk copied: %02X size: " + 2, maker[1]));
                } else if (maker[1] == (byte) 0xD9) {   // EOI
                    bos.write(maker);
                    Log.d("fuckshare", String.format("Chunk copied: %02X size: " + 2, maker[1]));
                    break;
                } else if (jpegSkippableChunks.contains(maker[1])) {
                    Utils.inputStreamRead(bis, lenBytes);
                    len = Utils.bigEndianBytesToLong(lenBytes) - 2;
                    Utils.inputStreamSkip(bis, len);
                    Log.d("fuckshare", String.format("Discord chunk: %02X size: " + (len + 4), maker[1]));
                } else if (maker[1] == (byte) 0xDA) {   // SOS
                    bos.write(maker);
                    // write all data
                    len = FileUtils.copy(bis, bos);
                    Log.d("fuckshare", "DA and following Chunks copied size: " + (len + 2));
                } else {
                    Utils.inputStreamRead(bis, lenBytes);
                    len = Utils.bigEndianBytesToLong(lenBytes) - 2;
                    bos.write(maker);
                    bos.write(lenBytes);
                    Utils.copy(bis, bos, len);
                    Log.d("fuckshare", String.format("Chunk copied: %02X size: " + (len + 4), maker[1]));
                }
            }
            bos.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/org/baiyu/fuckshare/exifhelper/JpegExifHelper.kt
package org.baiyu.fuckshare.exifhelper

import org.baiyu.fuckshare.Utils
import timber.log.Timber
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

class JpegExifHelper : ExifHelper {
    @Throws(IOException::class, ImageFormatException::class)
    override fun removeMetadata(inputStream: InputStream, outputStream: OutputStream) {
        val bis = inputStream.buffered()
        val bos = outputStream.buffered()

        val maker = ByteArray(2)
        val lenBytes = ByteArray(2)
        var chunkDataLength = 0L

        while (bis.available() > 0) {

            Utils.readNBytes(bis, maker)

            if (maker[0] != 0xFF.toByte()) {
                throw ImageFormatException()
            }

            when (maker[1]) {
                0xD8.toByte() -> {
                    bos.write(maker)
                    Timber.d("Copy chunk: %02X size: %d", maker[1], 2)
                }

                0xD9.toByte() -> {   // EOI
                    bos.write(maker)
                    Timber.d("Copy chunk: %02X size: %d", maker[1], 2)
                    break
                }

                0xDA.toByte() -> {   // SOS
                    bos.write(maker)
                    // write all data
                    chunkDataLength = bis.available().toLong()
                    Timber.d("Copy DA and following chunks size: %s", chunkDataLength + 2)
                    chunkDataLength -= bis.copyTo(bos)
                }

                in jpegSkippableChunks -> {
                    Utils.readNBytes(bis, lenBytes)
                    chunkDataLength = Utils.bigEndianBytesToLong(lenBytes) - 2
                    Timber.d("Discord chunk: %02X size: %d", maker[1], chunkDataLength + 4)
                    chunkDataLength -= Utils.skipNBytes(bis, chunkDataLength)
                }

                else -> {
                    Utils.readNBytes(bis, lenBytes)
                    chunkDataLength = Utils.bigEndianBytesToLong(lenBytes) - 2
                    bos.write(maker)
                    bos.write(lenBytes)
                    Timber.d("Copy chunk: %02X size: %d", maker[1], chunkDataLength + 4)
                    chunkDataLength -= Utils.copy(bis, bos, chunkDataLength)
                }
            }

            if (chunkDataLength != 0L) {
                throw ImageFormatException()
            }
        }
        bos.flush()
    }

    companion object {
        private val jpegSkippableChunks = setOf(
            0xE0.toByte(),
            0xE1.toByte(),
            0xE2.toByte(),
            0xE3.toByte(),
            0xE4.toByte(),
            0xE5.toByte(),
            0xE6.toByte(),
            0xE7.toByte(),
            0xE8.toByte(),
            0xE9.toByte(),
            0xEA.toByte(),
            0xEB.toByte(),
            0xEC.toByte(),
            0xED.toByte(),
            0xEE.toByte(),
            0xEF.toByte(),
            0xFE.toByte()
        )
    }
}